### PR TITLE
USGS collection 2 into WOfS for DEAfrica

### PR DESCRIPTION
**Pull request by eloise-b**
Main changes:
Edits to `filters.py`
- adding a C2 filter/mask, to do the cloud and cloud shadow masking
- uses the `pixel_quality` layer and from that layer `dilated_cloud` and `cloud_shadow`  

Edits to `virtual_product.py`
- if C2 scaling is used, also use the `woffles_c2`

Edits to `wofls.py`
- adding `woffles_c2` which uses `filter_c2`

### CODE REVIEW

**Comment by eloise-b** [comment]
@fangfy @omad @alexgleith

File: wofs/filters.py [hunk 1]
```diff
@@ -59,6 +59,62 @@ def pq_filter(pq):
...
+    #ipq = ~pq  # bitwise-not, e.g. flag cloudiness rather than cloudfree
+
+    masking = np.zeros(pq.shape, dtype=np.uint8)
+    #masking[(pq & C2_NODATA_BITS).astype(np.bool)] = constants.NO_DATA #retry the no data when it's not inverted
+    #masking[(ipq & C2_CLOUD_BITS)] += constants.MASKED_CLOUD
+    #masking[(ipq & C2_CLOUD_SHADOW_BITS)] += constants.MASKED_CLOUD_SHADOW #if you take dilate out, need astype bool
+    #masking[(pq & C2_CLOUD_BITS).astype(np.bool)] += constants.MASKED_CLOUD
+    #masking[((~pq & C2_CLEAR_BITS) | (pq & C2_CIRRUS_BITS)).astype(np.bool)] += constants.MASKED_CLOUD
+    #masking[((~pq & C2_CLEAR_BITS)).astype(np.bool)] += constants.MASKED_CLOUD
+    #masking[((pq & C2_DILATED_BITS) | (pq & C2_CIRRUS_BITS)).astype(np.bool)] += constants.MASKED_CLOUD
```
Comment (alexgleith) [inline wofs/filters.py:113]: Why are these all still there?

File: wofs/filters.py [hunk 2]
```diff
@@ -59,6 +59,62 @@ def pq_filter(pq):
     return masking
 
 
+C2_NODATA_BITS = 0x0001 # 0001 0th bit
+#C2_CLOUD_BITS = #0x000C # 1100 2nd and 3rd bit #try doing them as 2 checks instead of one
```
Comment (alexgleith) [inline wofs/filters.py:63]: Commented?

File: wofs/virtualproduct.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 from datacube.virtual import Transformation, Measurement
 from xarray import Dataset
 
-from wofs.wofls import woffles_ard
+from wofs.wofls import woffles_ard, woffles_c2
```
Comment (alexgleith) [inline wofs/virtualproduct.py:9]: I wonder if we should name it `woffles_usgs_c2` to be more specific.

File: wofs/virtualproduct.py [hunk 4]
```diff
@@ -59,6 +68,30 @@ def _load_dsm(self, gbox):
         return xr.Dataset(data_vars={'elevation': (('y', 'x'), dsm)}, coords=_to_xrds_coords(gbox),
                           attrs={'crs': gbox.crs})
 
+def scale_usgs_collection2(data):
+    return data.apply(scale_and_clip_dataarray, keep_attrs=True,
+                      scale_factor=0.275, add_offset=-2000, clip_range=(0, 10000))
```
Comment (alexgleith) [inline wofs/virtualproduct.py:73]: Might be nice to have a comment that references where we can check these values.

File: wofs/wofls.py [hunk 5]
```diff
@@ -56,6 +58,25 @@ def woffles_ard(ard, dsm):
 
     return water
 
+def woffles_c2(c2, dsm):
+    """Generate a Water Observation Feature Layer from ARD (NBART and FMASK) and surface elevation inputs."""
+    nbar_bands = spectral_bands(c2)
+    water = classifier.classify(nbar_bands) \
+        | eo_filter(c2) \
+        | c2_filter(c2.fmask)        
+        #| pq_filter(ard.pixel_quality)
```
Comment (alexgleith) [inline wofs/wofls.py:67]: Another comment to remove

**Review by alexgleith — CHANGES_REQUESTED** [review]
Some comments throughout.

One thing that I think you should change is to use less commits, so be deliberate about them.

I think this is a good tutorial on how to do that: https://dev.to/aman_singh/how-to-cleanup-commit-history-using-git-reset-46p3

**Comment by eloise-b** [comment]
@alexgleith I tried to clean up the commits, not sure what happened with it though! I also tidied up the parts that were commented out and updated `woffles_c2` to e `woffles_usgs_c2`

**Comment by alexgleith** [comment]
What I suggest you do is something like:

```
git reset --soft fe4dd7c197d7bfe2641545959249071b9340b3d2
```

Then do a new commit, so:

```
git commit -am "a nice message goes here"
```

And then push the results, either to a new branch (safe) or to this branch (`git push --force`, which is a little dangerous).

**Review by alexgleith — APPROVED** [review]

**Pull request merged**